iscsi.go

```diff
@@ -0,0 +1,475 @@
+}
+
+func logger() *slog.Logger {
+	return defaultLogger.Load()
+}
+
+type (
+	iscsiContext *C.struct_iscsi_context
+)
+
+type device struct {
+	Context      iscsiContext
+	targetName   string
+	targetPortal string
+	targetLun    int
+	details      ConnectionDetails
+}
+
+type ConnectionDetails struct {
+	InitiatorIQN string
+	TargetURL    string
+}
+
+func New(details ConnectionDetails) *device {
+	return &device{
+		details: details,
+	}
+}
+
+func (d *device) initializeContext() {
+	if d.Context != nil {
+		_ = C.iscsi_destroy_context(d.Context)
+		d.Context = nil
+		d.targetLun = 0
+		d.targetName = ""
+		d.targetPortal = ""
+	}
+	iqnStr := C.CString(d.details.InitiatorIQN)
+	defer C.free(unsafe.Pointer(iqnStr))
+	ctx := C.iscsi_create_context(iqnStr)
+	targetStr := C.CString(d.details.TargetURL)
+	defer C.free(unsafe.Pointer(targetStr))
+	url := C.iscsi_parse_full_url(ctx, targetStr)
+	_ = C.iscsi_set_targetname(ctx, &url.target[0])
+	defer C.iscsi_destroy_url(url)
+	d.Context = ctx
+	d.targetLun = int(url.lun)
+	d.targetName = C.GoString(&url.target[0])
+	d.targetPortal = C.GoString(&url.portal[0])
+	_ = C.iscsi_set_session_type(d.Context, C.ISCSI_SESSION_NORMAL)
+	_ = C.iscsi_set_header_digest(d.Context, C.ISCSI_HEADER_DIGEST_NONE_CRC32C)
+}
+
+func (d *device) Connect() error {
+	d.initializeContext()
+	return retry.Do(func() error {
+		portalStr := C.CString(d.targetPortal)
+		defer C.free(unsafe.Pointer(portalStr))
+		if retval := C.iscsi_full_connect_sync(d.Context, portalStr, C.int(d.targetLun)); retval != 0 {
+			errstr := C.iscsi_get_error(d.Context)
+			// reset the context before retrying.  it seems like some connection
+			// errors leave the context in an inconsistent state that makes it
+			// difficult to reuse
+			d.initializeContext()
+			return fmt.Errorf("iscsi_full_connect_sync: (%d) %s", retval, C.GoString(errstr))
+		}
+		return nil
+	}, retry.Attempts(20), retry.MaxDelay(500*time.Millisecond))
+}
+
+func (d *device) Reconnect() error {
+	if retval := C.iscsi_reconnect_sync(d.Context); retval != 0 {
+		if retval != 0 {
+			return fmt.Errorf("failed to reconnect with: %d", retval)
+		}
+	}
+	return nil
+}
+
+func (d *device) Disconnect() error {
+	defer C.iscsi_destroy_context(d.Context)
+	retval := C.iscsi_logout_sync(d.Context)
+	if retval != 0 {
+		return fmt.Errorf("failed to logout with: %d", retval)
+	}
+	return nil
+}
+
+type Capacity struct {
+	LBA       int
+	BlockSize int
+}
+
+func (d device) ReadCapacity10() (c Capacity, err error) {
+	task := C.iscsi_readcapacity10_sync(d.Context, 0, 0, 0)
+	defer func() {
+		if task != nil {
+			C.scsi_free_scsi_task(task)
+		}
+	}()
+	if task == nil || task.status != C.SCSI_STATUS_GOOD {
+		errstr := C.iscsi_get_error(d.Context)
+		return c, fmt.Errorf("iscsi_readcapacity10_sync: %s", C.GoString(errstr))
+	}
+	readcapacity, err := getReadCapacity10(*task)
+	if err != nil {
+		return c, err
+	}
+	c.BlockSize = int(readcapacity.block_size)
+	c.LBA = int(readcapacity.lba)
```

Comment:
Is `LBA` the number of blocks?

Reply:
Either way is cool with me. There's merit to matching the underlying implementation, so maybe just a comment would suffice. :+1: